### PR TITLE
spawn sync_endpoint_header_file goroutine out of createEndpoint func

### LINKED ISSUE #27371: Goroutines leaked: There are 100k+ go routines in the cilium agent and the cpu usage will be high
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Most of go routines' stack are 
```
 goroutine 3759128 [select]:
runtime.gopark(0x4a2a450, 0x0, 0x9, 0x18, 0x1)
	/usr/local/go/src/runtime/proc.go:363 +0xfd fp=0xc03ed5ac50 sp=0xc03ed5ac20 pc=0x44387d
          runtime.selectgo(0xc03ed5af80, 0xc03ed5ae1c, 0x0?, 0x0, 0x3?, 0x1)
 	/usr/local/go/src/runtime/select.go:328 +0xa49 fp=0xc03ed5ada8 sp=0xc03ed5ac50 pc=0x453ee9
    github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter(0xc028fa4d80)
 	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:209 +0x4ad fp=0xc03ed5afc0 sp=0xc03ed5ada8 pc=0x31f62cd
 github.com/cilium/cilium/pkg/trigger.NewTrigger.func1()
 	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:122 +0x2b fp=0xc03ed5afe0 sp=0xc03ed5afc0 pc=0x31f5a6b
 runtime.goexit()
 	/usr/local/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc03ed5afe8 sp=0xc03ed5afe0 pc=0x4761e1
 created by github.com/cilium/cilium/pkg/trigger.NewTrigger
	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:122 +0x30c
```

On this node, there is a pod which is allocated a wrong IP.  And it causes this issue. When creating the endpoints, it created a new goroutine. But it didnt shut down this goroutine before exiting when meeting an error.

```
 E0806 14:04:32.135739  103945 remote_runtime.go:105] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = failed to setup network for sandbox "69969e777f03433300fa750a6edeb61326c563e84cc8228901d773dae809ce3c": plugin type="cilium-cni" name="cilium" failed (add): unable to create endpoint: [PUT /endpoint/{id}][400] putEndpointIdInvalid  IP ipv4:10.42.170.215 is already in use
```
So kubelet/containerd will retry to create the endpoint again and again.
For the cilium agent, it will call createEndpoint func to create the endpoint.

```
func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, epTemplate *models.EndpointChangeRequest) (*endpoint.Endpoint, int, error) {
	...
	log.WithFields(logrus.Fields{
		"addressing":                 epTemplate.Addressing,
		...
	}).Info("Create endpoint request")

        // comment: in this func, it will create a new go routine
	ep, err := endpoint.NewEndpointFromChangeModel(d.ctx, owner, d, d.ipcache, d.l7Proxy, d.identityAllocator, epTemplate)
	if err != nil {
		return invalidDataError(ep, fmt.Errorf("unable to parse endpoint parameters: %s", err))
	}

	...

	for _, id := range checkIDs {
		oldEp, err := d.endpointManager.Lookup(id)
		if err != nil {
			return invalidDataError(ep, err)
		} else if oldEp != nil {
                         // comment: here it will not clean this go routine.
			return invalidDataError(ep, fmt.Errorf("IP %s is already in use", id))
		}
	}

	...
	// e.ID assigned here
	err = d.endpointManager.AddEndpoint(owner, ep, "Create endpoint from API PUT")
	if err != nil {
                // comment: there it starts to clean this go routine
		return d.errorDuringCreation(ep, fmt.Errorf("unable to insert endpoint into manager: %s", err))
	}
```
In the func `NewEndpointFromChangeModel`, it will call `func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, ifName string) *Endpoint`, and in this, it will call `initDNSHistoryTrigger` and create a go routine to do `syncEndpointHeaderFile` . 

```
func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, ifName string) *Endpoint {
	ep := &Endpoint{
		*
		allocator:        allocator,
		logLimiter:       logging.NewLimiter(10*time.Second, 3), // 1 log / 10 secs, burst of 3
		noTrackPort:      0,
	}

	ep.initDNSHistoryTrigger()
```
```
func (e *Endpoint) initDNSHistoryTrigger() {
	// Note: This can only fail if the trigger func is nil.
	var err error
	e.dnsHistoryTrigger, err = trigger.NewTrigger(trigger.Parameters{
		Name:        "sync_endpoint_header_file",
		MinInterval: 5 * time.Second,
		TriggerFunc: e.syncEndpointHeaderFile,
	})
	if err != nil {
		log.WithField(logfields.EndpointID, e.ID).WithError(err).Error("Failed to create the endpoint header file sync trigger")
	}
}
func NewTrigger(p Parameters) (*Trigger, error) {
	...

	go t.waiter()

	return t, nil
}
func (t *Trigger) waiter() {
	sleepTimer, sleepTimerDone := inctimer.New()
	defer sleepTimerDone()
	for {
		// keep critical section as small as possible
		t.mutex.Lock()
		triggerEnabled := t.trigger
		t.trigger = false
		t.mutex.Unlock()

		// run the trigger function
		if triggerEnabled {
		...
		}

		select {
		case <-t.wakeupChan:
		case <-sleepTimer.After(t.params.sleepInterval):

		case <-t.closeChan:
			shutdownFunc := t.params.ShutdownFunc
			if shutdownFunc != nil {
				shutdownFunc()
			}
			return
		}
	}
}
```
These go routines will try to get a lock at regular inervals. When the goroutines number is increasing to enough, it will cause that the lock competition in go-deadlock is fierce. The cilium agent cpu usage will be very high.
```
 sync.(*Mutex).Lock(0xc000011b90)
 	/usr/local/go/src/sync/mutex.go:90 +0x51 fp=0xc005bbf910 sp=0xc005bbf8e8 pc=0x4855f1
 github.com/sasha-s/go-deadlock.(*lockOrder).preLock(0xc000011b90, {0xc033be1ba0, 0x4, 0x32}, {0x42d9160, 0xc015c96fc0})
```

Questions:
1. Why we need to call `initDNSHistoryTrigger` in `func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, ifName string) *Endpoint `? I think this func 's main purpose is to assign the variables. 
Can we move this part into `func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.Endpoint, reason string) (err error) `?

2. In the patching endpoint, it will construct an intermediate variable by func `NewEndpointFromChangeModel`, it will create a gorotine, and it seems that this go routine is not shut down. So does patching endpoint cause the go routines leaked too?
```
func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Responder {
	...
	newEp, err2 := endpoint.NewEndpointFromChangeModel(h.d.ctx, h.d, h.d, h.d.ipcache, h.d.l7Proxy, h.d.identityAllocator, epTemplate)
```

3. I think we can have two optional patches. I think patch 1 is better. Will patch 1 involve some other issues?
patch 1: move initDNSHistoryTrigger into AddEndpoint
```
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 6273ece696..1d8d7ea6b1 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -466,8 +466,6 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
                noTrackPort:      0,
        }

-       ep.initDNSHistoryTrigger()
-
        ctx, cancel := context.WithCancel(context.Background())
        ep.aliveCancel = cancel
        ep.aliveCtx = ctx
@@ -479,7 +477,7 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
        return ep
 }

-func (e *Endpoint) initDNSHistoryTrigger() {
+func (e *Endpoint) InitDNSHistoryTrigger() {
        // Note: This can only fail if the trigger func is nil.
        var err error
        e.dnsHistoryTrigger, err = trigger.NewTrigger(trigger.Parameters{
@@ -817,7 +815,7 @@ func parseEndpoint(ctx context.Context, owner regeneration.Owner, policyGetter p
                return nil, fmt.Errorf("failed to parse restored endpoint: %s", err)
        }

-       ep.initDNSHistoryTrigger()
+       ep.InitDNSHistoryTrigger()

        // Validate the options that were parsed
        ep.SetDefaultOpts(ep.Options)
diff --git a/pkg/endpointmanager/manager.go b/pkg/endpointmanager/manager.go
index 0b13fbf2b3..313c0d0762 100644
--- a/pkg/endpointmanager/manager.go
+++ b/pkg/endpointmanager/manager.go
@@ -605,6 +605,7 @@ func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
        if ep.ID != 0 {
                return fmt.Errorf("Endpoint ID is already set to %d", ep.ID)
        }
+       ep.InitDNSHistoryTrigger()
        err = mgr.expose(ep)
        if err != nil {
                return err
```

patch 2: shutdown the goroutines if ep.DnsHistoryTrigger != nil 
```
diff --git a/daemon/cmd/endpoint.go b/daemon/cmd/endpoint.go
index 4fc5422726..875708d6c6 100644
--- a/daemon/cmd/endpoint.go
+++ b/daemon/cmd/endpoint.go
@@ -225,6 +225,9 @@ func invalidDataError(ep *endpoint.Endpoint, err error) (*endpoint.Endpoint, int
        ep.Logger(daemonSubsys).WithError(err).Warning("Creation of endpoint failed due to invalid data")
        if ep != nil {
                ep.SetState(endpoint.StateInvalid, "Invalid endpoint")
+               if ep.DnsHistoryTrigger != nil {
+                       ep.DnsHistoryTrigger.Shutdown()
+               }
        }
        return nil, PutEndpointIDInvalidCode, err
 }
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 6273ece696..f3da41b803 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -212,7 +212,7 @@ type Endpoint struct {

        // dnsHistoryTrigger is the trigger to write down the ep_config.h to make
        // sure that restores when DNS policy is in there are correct
-       dnsHistoryTrigger *trigger.Trigger
+       DnsHistoryTrigger *trigger.Trigger

        // state is the state the endpoint is in. See setState()
        state State
@@ -466,8 +466,6 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
                noTrackPort:      0,
        }

-       ep.initDNSHistoryTrigger()
-
        ctx, cancel := context.WithCancel(context.Background())
        ep.aliveCancel = cancel
        ep.aliveCtx = ctx
@@ -482,7 +480,7 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
 func (e *Endpoint) initDNSHistoryTrigger() {
        // Note: This can only fail if the trigger func is nil.
        var err error
-       e.dnsHistoryTrigger, err = trigger.NewTrigger(trigger.Parameters{
+       e.DnsHistoryTrigger, err = trigger.NewTrigger(trigger.Parameters{
                Name:        "sync_endpoint_header_file",
                MinInterval: 5 * time.Second,
                TriggerFunc: e.syncEndpointHeaderFile,
@@ -1149,8 +1147,8 @@ func (e *Endpoint) leaveLocked(proxyWaitGroup *completion.WaitGroup, conf Delete
        e.controllers.RemoveAll()
        e.cleanPolicySignals()

-       if e.dnsHistoryTrigger != nil {
-               e.dnsHistoryTrigger.Shutdown()
+       if e.DnsHistoryTrigger != nil {
+               e.DnsHistoryTrigger.Shutdown()
        }

        if e.ConntrackLocalLocked() {
@@ -2117,8 +2115,8 @@ func (e *Endpoint) syncEndpointHeaderFile(reasons []string) {
 // SyncEndpointHeaderFile triggers the header file sync to the ep_config.h
 // file. This includes updating the current DNS History information.
 func (e *Endpoint) SyncEndpointHeaderFile() {
-       if e.dnsHistoryTrigger != nil {
-               e.dnsHistoryTrigger.Trigger()
+       if e.DnsHistoryTrigger != nil {
+               e.DnsHistoryTrigger.Trigger()
        }
 }
```

### Cilium Version

1.13

### Kernel Version

5.15

### Kubernetes Version

1.18

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by sofat1989**
Fixes: #27371

```release-note
bug fix: sync_endpoint_header_file goroutines leaked if creating endpoints is failed because of invalidDataError
```

### CODE REVIEW

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR!

The fix makes sense, but I think just moving the init of the DNS trigger after this [line](https://github.com/cilium/cilium/blob/7530fd701ab15258d57e3731ea91bbd4a1fcf795/pkg/endpointmanager/manager.go#L618). If the endpoint creation fails, the cleanup of the DNS trigger happens in the endpoint cleanup function, `leaveLocked` which is called on all deletion paths.

File: daemon/cmd/endpoint.go [hunk 1]
```diff
@@ -469,6 +469,7 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 		}
 	}
 
+	ep.InitDNSHistoryTrigger()
```
Comment (tommyp1ckles) [inline daemon/cmd/endpoint.go:472]: If adding to the endpointmanager fails, should we still start this trigger?

**Comment by lmb** [comment]
I’ve marked this PR as draft while you work through the feedback in the PR above. When you’re ready to continue with review & testing of the PR, please click “Ready for Review” at the bottom of the page.

**Comment by aanm** [comment]
Hi @sofat1989 do you have an update on this issue?

**Comment by sayboras** [comment]
@sofat1989 friendly ping, not sure if you have any update or this is no longer the issue. Thanks.

**Pull request closed without merging**